metals/src/main/scala/scala/meta/internal/parsing/Trees.scala

```diff
@@ -0,0 +1,101 @@
+package scala.meta.internal.parsing
+
+import scala.collection.concurrent.TrieMap
+
+import scala.meta._
+import scala.meta.internal.metals.Buffers
+import scala.meta.internal.metals.BuildTargets
+import scala.meta.internal.mtags.MtagsEnrichments._
+import scala.meta.parsers.Parsed
+
+import org.eclipse.lsp4j.Diagnostic
+import org.eclipse.lsp4j.DiagnosticSeverity
+
+/**
+ * Manages parsing of Scala source files into Scalameta syntax trees.
+ *
+ * - provides the latest good Scalameta tree for a given source file
+ *   similar as `Buffers` provides the current text content.
+ * - produces diagnostics for syntax errors.
+ */
+final class Trees(
+    buildTargets: BuildTargets,
+    buffers: Buffers,
+    defaultDialect: Dialect = dialects.Scala213
+) {
+
+  private val trees = TrieMap.empty[AbsolutePath, Tree]
+
+  def get(path: AbsolutePath): Option[Tree] =
+    trees.get(path).orElse {
+      // Fallback to parse without caching result.
+      parse(path).flatMap { case (parsedSource, _) => parsedSource.toOption }
+    }
+
+  def didClose(fileUri: AbsolutePath): Unit = {
+    trees.remove(fileUri)
+  }
+
+  def didChange(path: AbsolutePath): List[Diagnostic] = {
+    parse(path) match {
```

Comment:
Maybe we can skip parsing scala3 code here, for now?

```scala
val dial = dialect(path)
if (dial.allowSignificantIndentation) {
  List()
} else {
  parse(path) match {
  // ...
```

Reply:
It's still useful to parse it as a lot of the functionalities will start to work even without any additional changes.